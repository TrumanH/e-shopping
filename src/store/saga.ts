import { categoriesSaga } from './categories/categories.saga';
import createSagaMiddleware from 'redux-saga';
import { all, call } from 'redux-saga/effects';

export function* rootSaga() {
    yield all([call(categoriesSaga)]);
  }

// create the saga middleware
export const sagaMiddleware = createSagaMiddleware();