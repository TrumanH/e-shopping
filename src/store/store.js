import { configureStore } from '@reduxjs/toolkit';
import userReducer from './user/user.slice';
import categoriesReducer from './categories/categories.slice';
import cartReducer from './cart/cart.slice';
// root reducer

export const store = configureStore({
  reducer: {
    user: userReducer,
    categories: categoriesReducer,
    cart: cartReducer,
  },
});