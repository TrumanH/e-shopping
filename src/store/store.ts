import userReducer from './user/user.slice';
import categoriesReducer from './categories/categories.slice';
import cartReducer from './cart/cart.slice';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { sagaMiddleware, rootSaga } from './saga';

const rootReducer = combineReducers({
  user: userReducer,
  categories: categoriesReducer,
  cart: cartReducer,
});

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  // blacklist: ['user'],
  whitelist: ['cart'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        // Ignore these action types
        ignoredActions: ['user/setUser', FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        // Ignore these field paths in all actions
        ignoredActionPaths: ['meta.arg', 'payload.timestamp'],
        // Ignore these paths in the state
        ignoredPaths: ['user.user'],
      },
    }).concat(sagaMiddleware),
  devTools: process.env.NODE_ENV !== 'production', // The Redux DevTools Extension is disabled for production
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

// then run the saga
sagaMiddleware.run(rootSaga)

export const persistor = persistStore(store);

// https://redux-toolkit.js.org/usage/usage-guide#use-with-redux-persist
