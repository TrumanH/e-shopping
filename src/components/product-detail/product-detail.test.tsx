
import ProductDetails from './product-detail.component';
import {render, screen} from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';
import {Provider} from "react-redux";
import {store} from "../../store/store";

describe('ProductDetails component', () => {
    const product = {
        id: 1,
        name: "hat name",
        title: "hat title",
        imageUrl: "https://i.ibb.co/ZYW3VTp/brown-brim.png",
        price: 10,
        category: "hats",
        desc: "description content",
        }

    const produceComponent = () => {
        return render(
          <BrowserRouter>
            <Provider store={store}>
              <ProductDetails product={product} />
            </Provider>
          </BrowserRouter>
          
        );
    }
    test('match the snapshot', ()=>{
        const { asFragment } = produceComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    test('render the ProductDetails component, get name, title and price etc.', ()=>{
        produceComponent();
        expect(screen.getByText(/hat title/i)).toBeInTheDocument();
        expect(screen.getByText(/Product name: hat name./i)).toBeInTheDocument();
        expect(screen.getByText(/Category: hats./i)).toBeInTheDocument();
        expect(screen.getByText(/10/i)).toBeInTheDocument();
        expect(screen.getByText(/ADD TO CART/i)).toBeInTheDocument();
        expect(screen.getByText(/Shipping fee: Free shipping./i)).toBeInTheDocument();
    });
});