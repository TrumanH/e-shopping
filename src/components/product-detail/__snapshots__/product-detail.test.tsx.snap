// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails component match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gKPRtg gnxzBh"
  >
    <div
      class="sc-iBYQkv kvqNyH"
    >
      hat title
    </div>
    <div
      class="sc-ftTHYK gbOShF"
    />
    <div
      class="sc-pyfCe fXLYXf"
    >
      <div
        class="sc-csuSiG hNmClS"
      >
        <div
          class="sc-eDWCr jXujlU"
        >
          $10
        </div>
        <div
          class="sc-bqWxrE idwknZ"
        >
          FREE SHIPPING
        </div>
      </div>
      <div
        class="sc-ksBlkl ktPCv"
      >
        Ready to dress up or down, this fashion product "hat name" is poping up now, it's been collected in our "hats" category.
      </div>
      <a
        class="sc-jrcTuL jHfhAy"
        href="#"
      >
        More information â†’ 
      </a>
      <div
        class="sc-hBxehG dMUjYu"
      >
        <div
          class="sc-fnGiBr hSazdw"
          color="black"
        />
        <div
          class="sc-fnGiBr epremR"
          color="blue"
        />
        <div
          class="sc-fnGiBr haXDRk"
          color="red"
        />
        <div
          class="sc-fnGiBr emRjSH"
          color="yellow"
        />
        <div
          class="sc-fnGiBr fmOIjQ"
          color="green"
        />
        <div
          class="sc-fnGiBr kmHHNN"
          color="brown"
        />
      </div>
    </div>
    <div
      class="sc-kDvujY bYVQOq"
    >
      <h3
        class="sc-fEXmlR bPsMWd"
      >
        product details
      </h3>
      <ul
        class="sc-bjfHbI gjJLzy"
      >
        <li
          class="sc-idXgbr gDreXT"
        >
          Poping up fashion item in our catetory hats collection.
        </li>
        <li
          class="sc-idXgbr gDreXT"
        >
          Category: hats.
        </li>
        <li
          class="sc-idXgbr gDreXT"
        >
          Product name: hat name.
        </li>
        <li
          class="sc-idXgbr gDreXT"
        >
          Shipping fee: Free shipping.
        </li>
      </ul>
    </div>
    <button
      class="sc-dkrFOg sc-eDvSVe bAenDU iJPJvM sc-ipEyDJ iDXrum"
    >
      ADD TO CART
    </button>
  </div>
</DocumentFragment>
`;
